server/service/src/sync/file_synchroniser.rs

```diff
@@ -0,0 +1,221 @@
+use chrono::{Duration, Utc};
+use rand::prelude::SliceRandom;
+use reqwest::multipart;
+use std::sync::Arc;
+use std::{cmp, io::Read};
+
+use repository::{
+    sync_file_reference_row::{
+        SyncFileReferenceRow, SyncFileReferenceRowRepository, SyncFileStatus,
+    },
+    RepositoryError,
+};
+
+use crate::{
+    service_provider::ServiceProvider,
+    static_files::{StaticFileCategory, StaticFileService},
+};
+
+use super::settings::SyncSettings;
+
+pub static MAX_UPLOAD_ATTEMPTS: i32 = 7 * 24; // 7 days * 24 hours Retry sending for up to for 1 week before giving up
+pub static RETRY_DELAY_MINUTES: i64 = 15; // Doubles each retry until MAX_RETRY_DELAY_MINUTES
+pub static MAX_RETRY_DELAY_MINUTES: i64 = 60; // 1 hour
+
+#[derive(Debug)]
+pub(crate) enum FileSyncError {
+    DatabaseError(RepositoryError),
+    CantFindFile(String),
+    StdIoError(std::io::Error),
+    ReqwestError(reqwest::Error),
+    UploadError(String), //TODO improve error handling (e.g. if it needs to wait to sync etc, maybe handle permanent vs temporary errors?)
+}
+
+impl From<RepositoryError> for FileSyncError {
+    fn from(error: RepositoryError) -> Self {
+        FileSyncError::DatabaseError(error)
+    }
+}
+
+pub struct FileSynchroniser {
+    settings: SyncSettings,
+    service_provider: Arc<ServiceProvider>,
+    static_file_service: Arc<StaticFileService>,
+    client: reqwest::Client,
+}
+
+impl FileSynchroniser {
+    pub(crate) fn new(
+        settings: SyncSettings,
+        service_provider: Arc<ServiceProvider>,
+        static_file_service: Arc<StaticFileService>,
+    ) -> Self {
+        Self {
+            settings,
+            service_provider,
+            static_file_service,
+            client: reqwest::Client::new(),
+        }
+    }
+
+    pub(crate) async fn sync(&self) -> Result<usize, FileSyncError> {
+        let ctx = self.service_provider.basic_context()?;
+
+        // Find any files that need to be uploaded
+        // Pick a file to upload
+        // Upload a file (In future this could be a chunk of data, instead of a whole file)
+        // Update the file record with the progress
+        // Yield to the runtime to check if we've received a pause signal
+
+        // Get any files that need to be sent to central server
+        let sync_file_repo = SyncFileReferenceRowRepository::new(&ctx.connection);
+        let files = sync_file_repo.find_all_to_upload()?;
+
+        // Pick a random file and upload it
+        // A random file is chosen in case one particular file is having problems, others should be able to continue
```

Comment:
Not sure random is needed now I've added a retry delay.
Previously had a bug were if it was trying to upload a file to central but central had lost the sync record (due to re-initialising I think) This meant file sync was stuck retrying forever even though it will never work..

Reply:
Yeah, was trying to create an opportunity for the `pause` message to get through and stop the uploading.
We could potentially cancel the inprogress upload process somehow, but that seems difficult and possibly problematic if a file takes more time than a single sync to upload...

Reply:
> if a file takes more time than a single sync to upload...

If the file takes longer then the interval/delay between syncs you mean ?

Reply:
Yes